### PR TITLE
[msl-out] Implement bounds checks for ImageLoad and ImageStore

### PR DESCRIPTION
**Pull request by jimblandy**
Posted early for silly reasons. ImageStore coming tomorrow. Comments welcome, of course.

### CODE REVIEW

**Review by kvark — APPROVED** [review]
Very nice!

**Comment by jimblandy** [comment]
just a rebase, comments not yet addressed

File: src/back/msl/writer.rs [hunk 1]
```diff
@@ -689,6 +731,191 @@ impl<W: Write> Writer<W> {
         Ok(())
     }
 
+    fn put_image_coordinate_limits(
+        &mut self,
+        image: Handle<crate::Expression>,
+        level: Option<LevelOfDetail>,
+        context: &ExpressionContext,
+    ) -> BackendResult {
+        self.put_image_size_query(image, level, "metal::uint", context)?;
```
[the remark this answers is not in the document]
Comment (jimblandy) [reply src/back/msl/writer.rs:740]: It's a bit exasperating that `uint` *must* be namespaced, while `int` *must not* be namespaced.

[the remark this answers is not in the document]
Comment (jimblandy) [reply src/back/msl/writer.rs:740]: I found a not-awful way to fix this, I think.

**Comment by jimblandy** [comment]
I addressed the `metal::uint` namespace issue in a separate PR, #1740.

**Comment by jimblandy** [comment]
Okay, I've looked over all the output one last time, added some minor cleanups, and rebased on master.

**Pull request merged**